File: public/packages/typescript/caching/abstractions/src/i-cache-serializer.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import type { D2Result } from "@dcsv-io/d2-result";

/**
 * Pluggable serialization seam for distributed caches. Local caches
 * store objects directly and do not need this. This package owns the
 * port only; a default JSON implementation is part of the
 * `@dcsv-io/d2-caching-distributed-redis` package surface. Provider-specific
 * impls may swap in MessagePack, Protobuf, etc., for size or perf wins.
 *
 * Impl failure codes (not defined here — reuse existing catalogs when
 * implementing): `COULD_NOT_BE_SERIALIZED` /
 * `COULD_NOT_BE_DESERIALIZED`.
 */
export interface ICacheSerializer {
  /**
   * Stable identifier for the serialization format (free string, e.g.
   * `"application/json"`). Allows mixed-serializer DLQs / archive
   * blobs to round-trip safely.
   */
  readonly contentType: string;

  /**
   * Serializes `value` to bytes.
   *
   * @returns `ok(bytes)`; failure with a serialize error code on
   *   serializer error.
   */
  serialize<T>(value: T): D2Result<Uint8Array>;

  /**
   * Deserializes `bytes` back to a value.
   *
   * @returns `ok(value)`; failure with a deserialize error code on
   *   deserializer error.
   */
  deserialize<T>(bytes: Uint8Array): D2Result<T>;
}

File: public/packages/dotnet/caching/abstractions/ICacheSerializer.cs
// -----------------------------------------------------------------------
// <copyright file="ICacheSerializer.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Caching;

using DcsvIo.D2.Result;

/// <summary>
/// Pluggable serialization seam for distributed caches. Local caches store
/// objects directly and don't need this. The default implementation in
/// the distributed lib is JSON; provider-specific impls may swap in
/// MessagePack, Protobuf, etc., for size or perf wins.
/// </summary>
public interface ICacheSerializer
{
    /// <summary>
    /// Gets a stable identifier for the serialization format (e.g.
    /// <c>"application/json"</c>). Allows mixed-serializer DLQs / archive
    /// blobs to round-trip safely.
    /// </summary>
    string ContentType { get; }

    /// <summary>
    /// Serializes <paramref name="value"/> to bytes.
    /// </summary>
    /// <typeparam name="T">Value type.</typeparam>
    /// <param name="value">Value to serialize.</param>
    /// <returns>
    /// <c>Ok(bytes)</c>; failure with <c>COULD_NOT_BE_SERIALIZED</c> on
    /// serializer error.
    /// </returns>
    D2Result<byte[]> Serialize<T>(T value);

    /// <summary>
    /// Deserializes <paramref name="bytes"/> back to a value.
    /// </summary>
    /// <typeparam name="T">Expected value type.</typeparam>
    /// <param name="bytes">Stored bytes.</param>
    /// <returns>
    /// <c>Ok(value)</c>; failure with <c>COULD_NOT_BE_DESERIALIZED</c> on
    /// deserializer error.
    /// </returns>
    D2Result<T> Deserialize<T>(byte[] bytes);
}
